base/commands/migration/stages.go

```diff
@@ -89,31 +89,64 @@ func (st *Stages) startStage(ctx context.Context) func(stage.Statuser) error {
 		if err != nil {
 			return err
 		}
-		if err := st.startQueue.Put(ctx, serialization.JSON(b)); err != nil {
+		if err = st.startQueue.Put(ctx, serialization.JSON(b)); err != nil {
+			return err
+		}
+		ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+		defer cancel()
+		if err = st.waitForStatus(ctx, time.Second, statusInProgress, statusComplete); err != nil {
 			return err
 		}
 		return nil
 	}
 }
 
 func (st *Stages) migrateStage(ctx context.Context) func(statuser stage.Statuser) error {
-	return func(status stage.Statuser) error {
-		for {
-			s, err := st.readStatus(ctx)
-			if err != nil {
-				return fmt.Errorf("reading status: %w", err)
+	return func(stage.Statuser) error {
+		return st.waitForStatus(ctx, 5*time.Second, statusComplete)
+	}
+}
+
+func (st *Stages) waitForStatus(ctx context.Context, waitInterval time.Duration, targetStatuses ...status) error {
+	timeoutErr := fmt.Errorf("migration could not be completed: reached timeout while reading status: " +
+		"please ensure that you are using Hazelcast's migration cluster distribution and your DMT config points to that cluster")
+	for {
+		if err := ctx.Err(); err != nil {
+			if errors.Is(err, context.DeadlineExceeded) {
+				return timeoutErr
+			} else {
+				return fmt.Errorf("migration failed: %w", err)
 			}
-			switch s {
-			case statusComplete:
-				return nil
-			case statusCanceled:
-				return clcerrors.ErrUserCancelled
-			case statusFailed:
-				return errors.New("migration failed")
+		}
+		s, err := st.readStatus(ctx)
+		if err != nil {
+			if errors.Is(err, context.DeadlineExceeded) {
+				return timeoutErr
 			}
-			time.Sleep(5 * time.Second)
+			return fmt.Errorf("reading status: %w", err)
+		}
+		switch s {
+		case statusComplete:
+			return nil
+		case statusCanceled:
+			return clcerrors.ErrUserCancelled
+		case statusFailed:
+			return errors.New("migration failed")
+		}
+		if expectationMet(targetStatuses, s) {
```

Comment:
```suggestion
		if slices.Contains(targetStatuses, s) {
```
With that change `expectationMet` function may be removed.

Reply:
https://github.com/hazelcast/hazelcast-commandline-client/pull/358/commits/6185781ceb7f389158faff58acdd44702df13304